File: epi/SearchRowColSortedMatrix.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;
public class SearchRowColSortedMatrix {
  @EpiTest(testDataFile = "search_row_col_sorted_matrix.tsv")

  public static boolean matrixSearch(List<List<Integer>> A, int x) {
//    //T O(M + N) S O(1)
//    int m = A.size(), n = A.get(0).size();
//    int r = m - 1, c = 0;
//
//    while(r >= 0 && c < n) {
//      int cur = A.get(r).get(c);
//      if(cur == x) {
//        return true;
//      } else if(cur > x) {
//        r--;
//      } else if(cur < x) {
//        c++;
//      }
//    }
//    return false;
    return epi.kt.SearchRowColSortedMatrix.INSTANCE.matrixSearch(A, x);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SearchRowColSortedMatrix.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/SearchRowColSortedMatrix.kt
package epi.kt

object SearchRowColSortedMatrix {
    fun matrixSearch(A: List<List<Int>>, x: Int) : Boolean {
        //T O(M + N) S O(1)
        //Start from top/right
        val m = A.size
        val n = A[0].size

        var i = 0
        var j = n - 1

        while(i < m && j >= 0) {
            val cur = A[i][j]
            when {
                cur == x -> {
                    return true
                }
                cur > x -> {
                    j--
                }
                cur < x -> {
                    i++
                }
            }
        }

        return false
    }
}